draft-ietf-taps-impl.md

```diff
@@ -1030,39 +1048,78 @@ Data Unit: Message
-must consume the lowest available (even or odd, depending on the side) stream number; this
-rule is relevant when lower numbers become available because Connection objects associated
-to the streams are closed.
+: Connection objects can be mapped to an SCTP association or a stream in an SCTP association.
+Mapping Connection objects to SCTP streams is called "stream mapping" and has additional requirements as follows.
+The following explanation assumes a client-server communication model.
+
+Stream mapping requires an association to already be in place between the client
+and the server, and it requires the server to understand that a new incoming stream
+should be represented as a new Connection Object by the Transport Services system.
+To signal the use of stream mapping, the server SHOULD use the adaptation layer indication
+parameter of LISTEN.SCTP and the client SHOULD use the adaptation layer indication
+parameter of CONNECT.SCTP, both with the "Stream Mapping" adaptation code point.
+Reception of an adaptation layer indication notification with the same adaptation code point value
+then indicates that the peer is capable of stream mapping.
+
+In the absence of an adaptation layer indication notification carrying a "Stream Mapping"
+adaptation code point, stream mapping can still be
+used if the application is aware that the peer application does not directly access SCTP streams.
+If, for example, a transport-protocol-agnostic client application tries to connect to a
+native SCTP server application that does not implement stream mapping, and assumes
+that a new Message sent on this Connection will look like the first Message of a new
+Connection on the server side, communication will fail because the
+SCTP-based application on the server side does not obtain a new Connection.
+As a counter-example, if both the client and server are
+written over a Transport Services system, and the server listens on all available protocols, stream
+mapping works without the application having to be aware of it (a code example that works
+without the adaptation layer notification, described in {{NEATPy}}, is available from the
+NEATPy repository mentioned in {{appendix-implementations}} at the time of writing).
+
+A new SCTP stream is created by sending an SCTP message with a new stream id. Thus, to
+implement stream mapping,
+the Transport Services system MUST provide a newly created Connection Object to the
+application upon the reception of such a message. The necessary semantics to
+implement a Transport Services system's `Close` and `Abort` primitives are provided
+by the stream reconfiguration (reset) procedure described in {{?RFC6525}}. This also
+allows to re-use a stream id after resetting ("closing") the stream.
+To implement this functionality, SCTP stream reconfiguration {{?RFC6525}} MUST be
+supported by both the client and the server side.
+
+To avoid head-of-line blocking, stream mapping SHOULD only be implemented
```

Comment:
I don't think I agree with this. A non-8260 SCTP still provides stream multiplexing with some HOLB avoidance properties, even if it's not as good. This is your question (3), which I think we agreed on?

Reply:
Confused. Question 2 was "close = stream reset", 3 was "I-DATA to avoid HOLB".
What is it you don't agree with? The "SHOULD only"?

I'm sceptical because I believe that small messages could get enqueued behind HUGE other messages, and there's also no per-stream flow control, at least not in the absence of I-DATA. A native SCTP application may know what it's doing, but a TAPS application would just use Connections and get a blockage surprise. So that doesn't seem good to me.

Reply:
Ahhh yes right, I looked at a response by @martinduke and he had changed the numbering. Tsk :)  First github turns the numbering into 1-8 for me, then this... sorry folks  :)

Reply:
So - what is it we have and don't have agreement about?  The text now says MUST for the "stream reset" SCTP extension, as it allows to implement the equivalent of CLOSE (we could be less restrictive, but I fear that it makes things more complex... we have to invent timeouts for this; also, note that this hasn't been tested with an implementation - the only implementation that we have of doing this uses "stream reset").  For I-DATA, it says SHOULD.

Reply:
I'm afraid we've QUIC-optimized our model; we have connection close and abort, and connection group abort. For SCTP, we'd want connection group close and abort, and an optional connection abort, IIUC

We could decide to add connection group close; or, we could just use the lack of a connection (stream) close in SCTP to simply signify it in the API by closing all the connections/streams.

Personally, I think we should have graceful close and abort for connection groups and write something in the QUIC mapping about how the former is handled.

Reply:
You say "the lack of a connection (stream) close in SCTP": but we do have stream reset, and it's practically the same. This is why we require it with a MUST, and this capability is negotiated with the peer by SCTP itself - so if we require it on one side, we have it. A stream reset is quite like a FIN.

So, I don't see why you'd need a group close for SCTP?  It may be a convenient thing to have, in general, sure, but I don't see what this has to do with SCTP in particular - being an API addition, I see this as separate from this PR, in fact.

Reply:
yes, you're right; it's a clean close. It's the stream abort that we're missing.

So in theory a streaming protocol has four possible ending operations: connection close, stream close, connection abort, stream abort. QUIC has 3 of these; SCTP has 2, 3 with RFC 6525; HTTP/2 has all four. In some protocols these are half-closes or aborts, or full. TAPS currently has stream close, stream abort, and connection abort.

So I'm going to file a separate issue on this, because it's bigger than the SCTP mapping.

Reply:
About stream abort: not quite, I think. Using a "stream abort" would mean that the TCP-like half-closed behavior wouldn't work - you couldn't send a stream reset and then expect to receive something on a stream. However, that isn't possible anyway, because SCTP streams are only unidirectional. And, there's only a guarantee of a notification by a returning "stream reset" because we define it in this PR.

I agree that this is bigger and should be separate from this PR... but what's your plan?   Do we really have a big problem here?  In the end, it's a matter of having the right semantics at the API level. This began with just picking the strictest possible semantics: no half-closed connections, calling "close" means that what the app already gave to the transport system for transmission before the call will still be sent, but the app cannot expect to send *or* receive anything after this call. This way, SCTP (which doesn't allow half-closed associations) is covered, and the behavior is "correct" for TCP as well. These semantics also work for SCTP streams with stream reset as above.

So... altogether, I thought we'll be safe if we just go strict.

Now, quite recently, the wording was changed a bit, because of the need to support TCP half-closed connections (which is implemented via the "Final" Message Property). I felt a bit uneasy about this change, but in the end I couldn't see how things would go wrong with the wording that we now have in the API draft.

My high-level view is that we still have this rather strict thing, and then there is the option to use "Final", with no guarantee of it working (because we don't know if we get a protocol that supports half-closed connections). But there may be a devil in the details on how we worded it... definitely worth another careful look.